velox/docs/functions/json.rst

```diff
@@ -2,6 +2,89 @@
+supported types. Casting from MAP is supported when the key type of the map
+is BOOLEAN, TINYINT, SMALLINT, INTEGER, BIGINT, REAL, DOUBLE, or VARCHAR
+and value type of the map is one of the supported types. Additionally,
+element types of ARRAY and MAP and field types of ROW are also allowed to
+be JSON. Behaviors of the casts are shown with the examples below:
+
+``SELECT CAST(NULL AS JSON); -- NULL``
+``SELECT CAST(1 AS JSON); -- JSON '1'``
+``SELECT CAST(9223372036854775807 AS JSON); -- JSON '9223372036854775807'``
+``SELECT CAST('abc' AS JSON); -- JSON '"abc"'``
+``SELECT CAST(true AS JSON); -- JSON 'true'``
+``SELECT CAST(1.234 AS JSON); -- JSON '1.234'``
+``SELECT CAST(ARRAY[1, 23, 456] AS JSON); -- JSON '[1,23,456]'``
+``SELECT CAST(ARRAY[1, NULL, 456] AS JSON); -- JSON '[1,null,456]'``
+``SELECT CAST(ARRAY[ARRAY[1, 23], ARRAY[456]] AS JSON); -- JSON '[[1,23],[456]]'``
+``SELECT CAST(MAP_FROM_ENTRIES(ARRAY[('k1', 1), ('k2', 23), ('k3', 456)]) AS JSON); -- JSON '{"k1":1,"k2":23,"k3":456}'``
+``SELECT CAST(CAST(ROW(123, 'abc', true, JSON '["a"]') AS ROW(v1 BIGINT, v2 VARCHAR, v3 BOOLEAN, v4 JSON)) AS JSON); -- JSON '[123,"abc",true,["a"]]'``
+
+Notice that casting from NULL to JSON is not straightforward. Casting from
+a standalone NULL will produce a SQL NULL instead of JSON 'null'. However,
+when casting from arrays or map containing NULLs, the produced JSON will
+have nulls in it.
+
+Another thing to be aware of is that when casting from ROW to JSON, the
+result is a JSON array rather than a JSON object. This is because positions
+are more important than names for rows in SQL.
+
+Finally, keep in mind that casting a VARCHAR string to JSON does not directly
+turn the original string into JSON type. Instead, it creates a JSON text
+representing the original string. This JSON text is different from the original
+string since it has characters escaped and extra double quotes added.
+
+Cast from JSON
+--------------
+Casting a JSON text to a supported type returns the value represented by this
+JSON text. The JSON text must represent a valid value of the type it is casted
+to, or an error will be thrown. Casting to BOOLEAN, TINYINT, SMALLINT, INTEGER,
+BIGINT, REAL, DOUBLE or VARCHAR is supported. Casting to ARRAY and MAP is
+supported when the element type of the array is one of the supported types, or
+when the key type of the map is BOOLEAN, TINYINT, SMALLINT, INTEGER, BIGINT,
+REAL, DOUBLE, or VARCHAR and value type of the map is one of the supported types.
+Behaviors of the casts are shown with the examples below:
+
+``SELECT CAST(JSON 'null' AS VARCHAR); -- NULL``
+``SELECT CAST(JSON '1' AS INTEGER); -- 1``
+``SELECT CAST(JSON '9223372036854775807' AS BIGINT); -- 9223372036854775807``
+``SELECT CAST(JSON '"abc"' AS VARCHAR); -- abc``
+``SELECT CAST(JSON 'true' AS BOOLEAN); -- true``
+``SELECT CAST(JSON '1.234' AS DOUBLE); -- 1.234``
+``SELECT CAST(JSON '[1,23,456]' AS ARRAY(INTEGER)); -- [1, 23, 456]``
```

Comment:
So Velox ROW becomes JSON ARRAY becomes Velox ARRAY?

Reply:
Actually Presto allows casting JSON ARRAY to Presto ROW (https://prestodb.io/docs/current/functions/json.html#cast-from-json), so we'll allow casting JSON ARRAY to Velox ROW as well. That work is not finished yet, so I plan to add the relevant examples to this document after I commit that PR.